drivers/firmware/nrf_ironside/call.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/cache.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/firmware/nrf_ironside/call.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/barrier.h>
+#include <zephyr/sys/math_extras.h>
+#include <zephyr/sys/util.h>
+
+#define DT_DRV_COMPAT nordic_ironside_call
+
+#define SHM_NODE     DT_INST_PHANDLE(0, memory_region)
+#define NUM_BUFS     (DT_REG_SIZE(SHM_NODE) / sizeof(struct ironside_call_buf))
+#define ALL_BUF_BITS BIT_MASK(NUM_BUFS)
+
+/* Note: this area is already zero-initialized at reset time. */
+static struct ironside_call_buf *const bufs = (void *)DT_REG_ADDR(SHM_NODE);
+
+#if defined(CONFIG_DCACHE_LINE_SIZE)
+BUILD_ASSERT((DT_REG_ADDR(SHM_NODE) % CONFIG_DCACHE_LINE_SIZE) == 0);
+BUILD_ASSERT((sizeof(struct ironside_call_buf) % CONFIG_DCACHE_LINE_SIZE) == 0);
+#endif
+
+static const struct mbox_dt_spec mbox_rx = MBOX_DT_SPEC_INST_GET(0, rx);
+static const struct mbox_dt_spec mbox_tx = MBOX_DT_SPEC_INST_GET(0, tx);
+
+K_EVENT_DEFINE(alloc_evts);
+K_EVENT_DEFINE(rsp_evts);
+
+static void ironside_call_rsp(const struct device *dev, mbox_channel_id_t channel_id,
+			      void *user_data, struct mbox_msg *data)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel_id);
+	ARG_UNUSED(user_data);
+	ARG_UNUSED(data);
+
+	struct ironside_call_buf *buf;
+	uint32_t rsp_buf_bits = 0;
+	const uint32_t skip_buf_bits = k_event_test(&rsp_evts, ALL_BUF_BITS);
```

Comment:
I found it a bit hard to see how this is not race prone in regards to cache invalidating an 'in progress' buffer below. Can you leave a comment that we are inside the irq handler for the mbox and will only at the end post events so `skip_buf_bits` will remain valid throughout?

It could miss a new response because it might skip something that meanwhile became a valid pending request, but that is fine as this response handler will be called again.

Reply:
Added a longer comment.